### PR TITLE
[Core]Support compiled class cache in CodeGenUtils

### LINKED ISSUE #2955: [Feature] Support compiled class cache in CodeGenUtils
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/incubator-paimon/issues) and found nothing similar.

### Motivation

screenshot-20240306-171507

When we have a table with thousands of columns, the operation to generate a RecordEqualiser class will consume lots of resources. Especially when we are creating multiple Readers from multiple Table instances at the same time, this is a real case when reading the paimon table using Starrocks as external tables. By reusing the generated class, we can avoid this problem.

### Solution

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [X] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by xiangyuf**
### Purpose

Linked issue: close #2955 

Reuse the generated class in CodeGeneratorImpl to improve performance.

### Tests

org.apache.paimon.codegen.CodeGeneratorCacheTest

### CODE REVIEW

**Comment by xiangyuf** [comment]
Hi @JingsongLi  , PTAL when you have time. Thx.

**Comment by xiangyuf** [comment]
@FangYongs @liming30 Hi, would u kindly help review this pr?

File: paimon-codegen/src/main/java/org/apache/paimon/codegen/CodeGeneratorImpl.java [hunk 1]
```diff
@@ -68,4 +148,40 @@ private SortSpec getAscendingSortSpec(int[] sortFields) {
         }
         return builder.build();
     }
+
+    /** Class to use as key for the {@link #GENERATED_CLASS_CACHE}. */
+    static class ClassKey {
+        private final String name;
```
Comment (JingsongLi) [inline paimon-codegen/src/main/java/org/apache/paimon/codegen/CodeGeneratorImpl.java:154]: Should contains interface class, for example, `RecordEqualiser.class`.

Please add test for this.

**Review by JingsongLi — CHANGES_REQUESTED** [review]
Can you just add cache to `CodeGenUtils`? In that way, we can cache generated Class.

**Comment by xiangyuf** [comment]
> Can you just add cache to `CodeGenUtils`? In that way, we can cache generated Class.

@JingsongLi Do you mean cache the instance directly?

**Comment by JingsongLi** [comment]
> > Can you just add cache to `CodeGenUtils`? In that way, we can cache generated Class.
> 
> @JingsongLi Do you mean cache the instance directly?

I mea compiled `Class<T>`.

**Comment by JingsongLi** [comment]
Please fix compile

**Comment by JingsongLi** [comment]
merged in 4ec5806428a8d3ddbb0ba8521b8971f0e9de7b17

Looks like your email (fengxiangyu.fxy@bytedance.com) is not bind to your github id.

**Pull request closed without merging**